pkg/controllers/leaderworkerset_controller.go

```diff
@@ -200,6 +181,146 @@ func SetupIndexes(indexer client.FieldIndexer) error {
+// we didn't consider rollout strategy type here since we only support rollingUpdate now,
+// once we have more policies, we should update the logics here.
+// Possible scenarios:
+//   - When sts is under creation, partition is always 0 because pods are created in parallelism, rolling update is not fit here.
+//   - When sts is rolling update, the partition will start from the last index to the 0 index step by maxUnavailable.
+//   - When sts is rolling update, replicas increases, we'll delay the rolling update until scaling ended,
+//     Partition will not change, new replicas will apply with new template to avoid recreation.
+//   - When sts is rolling update, replicas decreases, the partition will not change until newReplicas < Partition,
+//     Partition will be reset to newReplicas.
+//   - When sts is ready to roll update and replica increases the same time, we'll delay the rolling update until scaling ended.
+//   - When sts is ready to roll update and replica decreases the same time, we'll start the rolling update together with scaling.
+//
+// When all set, the partition should always be zero.
+func (r *LeaderWorkerSetReconciler) rollPartition(ctx context.Context, lws *leaderworkerset.LeaderWorkerSet) (int32, error) {
+	sts := &appsv1.StatefulSet{}
+	err := r.Get(ctx, types.NamespacedName{Name: lws.Name, Namespace: lws.Namespace}, sts)
+	if err != nil {
+		// If sts not created yet, all partitions should be updated.
+		if apierrors.IsNotFound(err) {
+			return 0, nil
+		}
+		return 0, err
+	}
+
+	partition := *sts.Spec.UpdateStrategy.RollingUpdate.Partition
+	replicas := *lws.Spec.Replicas
+	rollingStep, err := intstr.GetValueFromIntOrPercent(&lws.Spec.RolloutStrategy.RollingUpdateConfiguration.MaxUnavailable, int(replicas), false)
+	if err != nil {
+		return 0, err
+	}
+
+	// Indicates a new rolling update here.
+	if templateUpdated(sts, lws) {
+		// Processing scaling up/down first prior to rolling update.
+		// When scaling up, update the Partition to the statefulset replicas
+		// When scaling down, update the Partition to the new replicas.
+		if replicasUpdated(sts, lws) {
+			return min(*lws.Spec.Replicas, *sts.Spec.Replicas), nil
+		} else {
+			return replicas - int32(rollingStep), nil
+		}
+	}
+
+	// Replicas changes during rolling update.
+	if replicasUpdated(sts, lws) {
+		return min(partition, *lws.Spec.Replicas), nil
+	}
+
+	// Otherwise always return 0.
+	if partition == 0 {
+		return 0, nil
+	}
+
+	stsSelector := client.MatchingLabels(map[string]string{
+		leaderworkerset.SetNameLabelKey: lws.Name,
+	})
+	var stsList appsv1.StatefulSetList
+	if err := r.List(ctx, &stsList, stsSelector, client.InNamespace(lws.Namespace)); err != nil {
+		return 0, err
+	}
+
+	var finishedPart int32
+	templateHash := utils.LeaderWorkerTemplateHash(lws)
+
+stop:
+	for i := replicas - 1; i >= 0; i-- {
+		for j := len(stsList.Items) - 1; j >= 0; j-- {
```

Comment:
This nested loop is expensive, how about split them into two linear loops: 

loop 1 iterates over the items and places them in a list based on their indexes (remember each item has an index).
loop 2 iterate over the indexed list in reverse order to check which ones are ready.

Reply:
Let me follow this next for performance issue.

Reply:
Thanks, the approach of splitting into two loops is more intuitive too and easier to follow.

Reply:
> Let me follow this next for performance issue.

In this PR, right?

Reply:
yes, I'm finishing the last part.